src/relay/backend/aot/annotate_used_memory.cc

```diff
@@ -0,0 +1,156 @@
+
+#include "../../transforms/device_aware_visitors.h"
+#include "../manifest_lifetimes.h"
+
+namespace tvm {
+namespace relay {
+namespace backend {
+namespace aot {
+
+/*!
+ * \brief Annotates the memory usage of each primitive function by analysing the liveness
+ * of the input/output tensors at the function callsite and calculating the total amount of
+ * memory these tensors require.
+ */
+class AnnotateUsedMemoryMutator : public transform::DeviceAwareExprMutator {
+ public:
+  AnnotateUsedMemoryMutator(const IRModule& module, const transform::ControlFlowGraph& cfg,
+                            const transform::LivenessAnalysis& lva)
+      : DeviceAwareExprMutator(module), control_flow_graph_(cfg), liveness_(lva) {}
+
+  /*!
+   * \brief Get the memory required for a primitive Relay function by calculating the total
+   * bytes of the live tensors at the callsite of the function.
+   *
+   * \param live_tensors The tensors that are live when the function is called.
+   * \return int The total number of bytes a function requires.
+   */
+  int GetMemoryUsage(const transform::VarSet& live_tensors) {
+    Array<Type> types_stack = {};
+    int memory_usage = 0;
+
+    for (const Var& var : live_tensors) {
+      Type var_type = var->checked_type();
+      ICHECK(var_type.defined()) << "InferTypes pass should be run before AnnotateUsedMemory pass.";
+      types_stack.push_back(var_type);
+    }
+
+    while (!types_stack.empty()) {
+      Type current_type = types_stack.back();
+      types_stack.pop_back();
+
+      if (const auto* tt_node = current_type.as<TupleTypeNode>()) {
+        for (const Type& type : tt_node->fields) {
+          types_stack.push_back(type);
+        }
+        continue;
+      } else if (const auto* ft_node = current_type.as<FuncTypeNode>()) {
+        types_stack.push_back(ft_node->ret_type);
+        continue;
+      }
+
+      const auto* tt_node = current_type.as<TensorTypeNode>();
+      ICHECK(tt_node) << "Expected TensorTypeNode but was " << current_type->GetTypeKey();
+      int total_tensor_bytes = GetTensorBytes(tt_node);
+      memory_usage += total_tensor_bytes;
+    }
+    return memory_usage;
+  }
+
+  /*!
+   * \brief Get the number of bytes a tensor requires.
+   *
+   * \param tensor_type_node The checked type of the tensor.
+   * \return int The number of bytes required.
+   */
+  int GetTensorBytes(const TensorTypeNode* tensor_type_node) {
+    PrimExpr size = tensor_type_node->Size();
+    const auto* size_int_imm = size.as<IntImmNode>();
+    ICHECK(size_int_imm) << "Expected tensor size to be an IntImmNode but was "
+                         << size->GetTypeKey();
+
+    int total_size = size_int_imm->value;
+    int dtype_bytes = tensor_type_node->dtype.bytes();
+    return total_size * dtype_bytes;
+  }
+
+  Expr PostVisitLet_(const LetNode* pre_let_node, const LetNode* post_let_node) override {
+    if (const auto* func_node = pre_let_node->value.as<FunctionNode>()) {
```

Comment:
is this matching a lowered function call? would be great if we could comment the expected construct here.

Reply:
yeah I am also confused, since (at least from IR perspective) this binding is not a call to a (primitive?) function but just binding the function to a var. I would have expected `pre_let_node->value.as<CallNode>()` with some check that the op in the call is a func. Maybe I am missing some context for the IR form at this point? definitely would like a comment

Reply:
Great point thanks, in hindsight we should definitely be visiting the callsite rather than the function. Previously I thought I could get all the information needed just visiting the function, since checking if the call node op is a function is a little more complex due to the let bindings getting in the way. But as you both rightly mention this isn't the correct way to do it